### DISCUSSION TITLE
ability to override / prioritize some stdlib stubs

### DISCUSSION THREAD

**Discussion by Josverl** [microsoft/pyright#2934] [Q&A]
I would like the ability to override the stubs for only a few (7-8) stdlib modules.
The rationale is that I'm working to provide static type checking for [MicroPython](https://micropython.org) , which provides a sub-set of Python 3.

I maintain a [collection of MicroPython stubfiles](https://github.com/Josverl/micropython-stubs#micropython-stubs) for the various different hardware ports and board variants, and they work quite well for type checking, with the exception of some of the stdlib modules.

For a part this can be accommodated in code as MicroPython also allows some ( not all) modules to be loaded with a `u` prefix, 
so while this works for 
``` python 
import utime as time 

time.sleep(1)           # sleep for 1 second
time.sleep_ms(500)      # sleep for 500 milliseconds - MicroPython extension
time.sleep_us(10)       # sleep for 10 microseconds - MicroPython extension
start = time.ticks_ms() # get millisecond counter - MicroPython extension
``` 
the below results in `is not a known member of` errors, even though there is a MicroPython specific `time.pyi` stub in the Search Paths.
``` python 
import time 

time.sleep(1)           # sleep for 1 second
time.sleep_ms(500)      # sleep for 500 milliseconds - MicroPython extension
time.sleep_us(10)       # sleep for 10 microseconds - MicroPython extension
start = time.ticks_ms() # get millisecond counter - MicroPython extension
``` 
However this approach is not possible for all modules, `builtins`, `gc`, and also there is a lot of MicroPython code that uses the standard names for reasons of compatibility or code style.

If possible I would like to be able to offer support for that.
I know it is possible to point to a local copy of typeshed using the 'typeshedPath' , but that would require me to maintain that copy of some 500 files or at least 70-100 in order to override (7-8 stubs](https://docs.micropython.org/en/latest/genrst/index.html). 

**Question:**
Is there a way to prioritize (a folder of) stub files over the stubs in `pyright\dist\typeshed-fallback\stdlib` ?
or is there another way to solve this without the need to duplicate the fallback\stdlib ?

**Comment by JelleZijlstra** [comment]
For what it's worth I'd be supportive of storing MicroPython stubs in typeshed, using something like `platform.python_implementation` (https://docs.python.org/3.10/library/platform.html#platform.python_implementation - why isn't micropython there?).

That would be a longer-term project though and require coordination among type checkers.

**Reply by erictraut** [reply]
@Josverl, out of curiosity, how different are these stubs from the current typeshed stubs? Do they differ in just a few minor ways (e.g. a small percentage of methods or classes are not present)? Or is it significantly different (most of the stdlib modules are missing, and many classes and methods in the other modules are different in significant ways)?

For context, pyright currently resolves imports in the following order:
1. stdlib typeshed (from typeshedPath if specified, or from the bundled typeshed stubs otherwise)
2. local stub (if stubPath is specified)
3. execution environment root directory
4. extraPath for execution environment
5. stub packages installed in selected python environment
6. normal (non-stub) packages installed in selected python environment
7. stubs bundled with pylance (not present in pyright)
8. third-party typeshed (from typeshedPath if specified, or from the bundled typeshed stubs otherwise)

Note that this deviates slightly from PEP 561. If we were to follow the guidance in this PEP, we'd need to move 1 down to just above 8. We had reasons to deviate from PEP 561 in the past, but those reasons are less important today than when we first made that decision. I think we should consider making the change to adhere to PEP 561. This is one potential way to address your use case. This change could negatively impact pylance use cases (e.g. completion suggestions within Jupyter notebooks), so we'll need to discuss and perhaps do some experimentation before committing to such a change.

Other potential solutions include:
1. Add a way to specify another path that is searched prior to 1 in the list above. (I don't like this solution because it involves adding yet another configuration knob that needs to be documented, maintained, etc. So I don't think we'd be inclined to do this.)
2. As Jelle suggested, expand the existing typeshed stubs to cover MicroPython through the use of conditional statements. This would require adding support for static evaluation of `python_implementation` in all current Python type checkers — and probably a small amendment to PEP 484 to specify this functionality.

I'll bring this up for discussion with the rest of the pylance team, and we'll get back to you.

**Reply by Josverl** [reply]
I think about a few different types of modules/stubs 
1) MicroPython language  - MicroPython [follows Python 3.4 with some 3.5+ aspects](https://docs.micropython.org/en/latest/genrst/index.html)
2) MicroPython core / stdlib  - limited to only a few essential modules. 
3) Hardware specific modules ;: ( pyb , machine, esp , esp32, _rp2 ) low level modules implemented in C / C++
4) MicroPython Frozen Modules:  port and board specific hardware modules , implemented in Python and cross-compiled and included in the firmware. Modules are source from micropython-lib ( or a '{family}-lib') 
5) Other micropython-lib modules are available on demand through PyPi [https://docs.micropython.org/en/latest/reference/glossary.html#term-micropython-lib]

@JelleZijlstra : good question as that is one of the many places I looked before starting to try rolling my own.
since then I may have found a few possible reasons:
1.  There are different open source and commercial `families` of MicroPython : Micropython , Pycopy , CircuitPython, Pycom to name a few 
2.  There are different variant and ports ( unix , windows and many different hardware ports ), and a lot of them heave their own hardware specific interfaces as modules, but also sometimes also as changes to a class or method of the 'mpy-stdlib'
3. based on my experience thus far I think the `platform.python_implementation()` would need to be structured something like :  `{firmware family}[-{port}[-{board}[-{variant}]]]`, or split out acriss multiple attributes. That indeed would need typecheckers and IDEs to agree on such a convention
4. For the Micropython family there are no stubs other than a few individually & manual maintained stubs. 
CircuitPython does have a stub package on PyPi ( circuitpython-stubs ) but that only contains stubs for hardware specific modules, not for the core language nor for the 'mpy-stdlib'

Im trying to get the stubs to PyPi , but want to make sure I have a decent structure, quality and usability in place.
Naturally I'm not the only one trying to close this gap https://github.com/hlovatt/PyBoardTypeshed

**Reply by Josverl** [reply]
@erictraut 

> out of curiosity, how different are these stubs from the current typeshed stubs? Do they differ in just a few minor ways (e.g. a small percentage of methods or classes are not present)? Or is it significantly different (most of the stdlib modules are missing, and many classes and methods in the other modules are different in significant ways)?


I would say :
- most of the stdlib functionality is not implemented , but personally I would not be bothered that my IDE does not flag that Pandas is not  availably on my rp2 microcontroller

Some of the functions / classes / methods are different on the stdlib sub-set 
 - where MicroPython does less/different than CPython this is documented in detail : see [MicroPython differences from CPython](https://docs.micropython.org/en/latest/genrst/index.html)
 - where MicroPython does more as in the `time.sleep_ms(500)` example above, this is not so clearly documented 

Im still building the test & typechecking suite for that to get a measurable view on what works and what not.
and these stdlib errors are the among the hard ones to solve.

I would estimate that the stdlib space to be less than <= 7 modules 
with approx 2-6 changes per module compared to make it not provide false negatives , not throw errors / warnings for code that runs well on a micropython port / board

**Reply by erictraut** [reply]
@Josverl, I discussed this issue with the rest of the pylance crew, and we decided that we're going to make the change to the import resolution ordering. This will bring us in compliance with PEP 561 and it will address your issue.

We're a little scared that this change may negatively affect some pyright and pylance users, so we're going to be careful in how we roll it out. It will first appear in pyright sometime later this week. Then we'll see if pyright users have any feedback (positive or negative). If everything looks good, the change will appear in some future version of pylance (perhaps next week). We need to be more careful with pylance because the user base is quite large, so there's a much bigger impact if we introduce a regression.

I've filed the following issue to track this change: https://github.com/microsoft/pyright/issues/2937.

**Reply by Josverl** [reply]
@erictraut, It makes perfect sense to do this carefully and gradually to make sure that the experience in VSCode / Pylance is not negatively impacted.
With your proposal I see a clear and good path forward as I was also looking at PEP 561 type-only packages for distribution.
